In [1]:
import requests
import json
import logging
import pandas as pd

In [3]:
logging.basicConfig(filename='GP_logs.log', level=logging.INFO, format='%(asctime)s %(message)s', datefmt='%Y-%m-%d %H:%M:%S', force=True)

In [4]:
headers = {
    'X-Api-App-Id': SJ_APP_SECRET,
    'User-Agent': 'User-Agent'
}

In [5]:
logging.info('отправляем первый запрос к апи суперджоб города')

In [6]:
towns = requests.get('https://api.superjob.ru/2.0/towns/?', headers=headers)

In [7]:
towns.json()

{'objects': [{'id': 4,
   'id_region': 46,
   'id_country': 1,
   'title': 'Москва',
   'title_eng': 'Moscow'},
  {'id': 14,
   'id_region': 41,
   'id_country': 1,
   'title': 'Санкт-Петербург',
   'title_eng': 'Saint-Petersburg'},
  {'id': 13,
   'id_region': 50,
   'id_country': 1,
   'title': 'Новосибирск',
   'title_eng': 'Novosibirsk'},
  {'id': 33,
   'id_region': 65,
   'id_country': 1,
   'title': 'Екатеринбург',
   'title_eng': 'Ekaterinburg'},
  {'id': 12,
   'id_region': 22,
   'id_country': 1,
   'title': 'Нижний Новгород',
   'title_eng': 'Nizhnii Novgorod'},
  {'id': 73,
   'id_region': 60,
   'id_country': 1,
   'title': 'Ростов-на-Дону',
   'title_eng': 'Rostov-na-Donu'},
  {'id': 56,
   'id_region': 8,
   'id_country': 1,
   'title': 'Хабаровск',
   'title_eng': 'Khabarovsk'},
  {'id': 55,
   'id_region': 92,
   'id_country': 1,
   'title': 'Казань',
   'title_eng': "Kazan'"},
  {'id': 106,
   'id_region': 75,
   'id_country': 1,
   'title': 'Челябинск',
   'title_eng

In [8]:
logging.info('города успешно')

In [9]:
logging.info('отправляем первый запрос к апи суперджоб каталоги')

In [10]:
catalogue = requests.get('https://api.superjob.ru/2.0/catalogues/?', headers=headers)

In [11]:
catalogue.json()

[{'title_rus': 'IT, Интернет, связь, телеком',
  'url_rus': 'it-internet-svyaz-telekom',
  'title': 'IT, Интернет, связь, телеком',
  'title_trimmed': 'IT, Интернет, связь,...',
  'key': 33,
  'positions': [{'title_rus': 'AI',
    'url_rus': 'ai',
    'title': 'AI',
    'id_parent': 33,
    'key': 651},
   {'title_rus': 'CRM-системы',
    'url_rus': 'crm-sistemy',
    'title': 'CRM-системы',
    'id_parent': 33,
    'key': 603},
   {'title_rus': 'Data Science',
    'url_rus': 'data-science',
    'title': 'Data Science',
    'id_parent': 33,
    'key': 627},
   {'title_rus': 'DevOps',
    'url_rus': 'devops',
    'title': 'DevOps',
    'id_parent': 33,
    'key': 628},
   {'title_rus': 'SRE',
    'url_rus': 'sre',
    'title': 'SRE',
    'id_parent': 33,
    'key': 629},
   {'title_rus': 'Web-верстка',
    'url_rus': 'web-verstka',
    'title': 'Web-верстка',
    'id_parent': 33,
    'key': 36},
   {'title_rus': 'Администрирование баз данных',
    'url_rus': 'administrirovanie-baz-danny

In [12]:
logging.info('каталоги успешно')

Более-менее разобрались, теперь будем это делать с помощью функции

In [13]:
def get_request(url):
    logging.info(f'отправляем запрос к {url}')
    try:
        request = requests.get(url, headers=headers)
        logging.info(f'{url} успех')
        return request.json()
    except Exception as e:
        logging.error(f'{url} ошибка {e}')

In [14]:
get_request('https://api.superjob.ru/2.0/catalogues/?')

[{'title_rus': 'IT, Интернет, связь, телеком',
  'url_rus': 'it-internet-svyaz-telekom',
  'title': 'IT, Интернет, связь, телеком',
  'title_trimmed': 'IT, Интернет, связь,...',
  'key': 33,
  'positions': [{'title_rus': 'AI',
    'url_rus': 'ai',
    'title': 'AI',
    'id_parent': 33,
    'key': 651},
   {'title_rus': 'CRM-системы',
    'url_rus': 'crm-sistemy',
    'title': 'CRM-системы',
    'id_parent': 33,
    'key': 603},
   {'title_rus': 'Data Science',
    'url_rus': 'data-science',
    'title': 'Data Science',
    'id_parent': 33,
    'key': 627},
   {'title_rus': 'DevOps',
    'url_rus': 'devops',
    'title': 'DevOps',
    'id_parent': 33,
    'key': 628},
   {'title_rus': 'SRE',
    'url_rus': 'sre',
    'title': 'SRE',
    'id_parent': 33,
    'key': 629},
   {'title_rus': 'Web-верстка',
    'url_rus': 'web-verstka',
    'title': 'Web-верстка',
    'id_parent': 33,
    'key': 36},
   {'title_rus': 'Администрирование баз данных',
    'url_rus': 'administrirovanie-baz-danny

Давайте теперь разберемся, какие параметры нам нужны

In [15]:
for c in catalogue.json():
    print(c['title_rus'])

IT, Интернет, связь, телеком
Административная работа, секретариат, АХО
Банки, инвестиции, лизинг
Безопасность, службы охраны
Бухгалтерия, финансы, аудит
Государственная служба
Дизайн
Домашний персонал
Закупки, снабжение
Искусство, культура, развлечения
Кадры, управление персоналом
Консалтинг, стратегическое развитие
Маркетинг, реклама, PR
Медицина, фармацевтика, ветеринария
Наука, образование, повышение квалификации
Некоммерческие организации, волонтерство
Обмен персоналом
Продажи
Промышленность, производство
Рабочий персонал
Сельское хозяйство
Службы доставки
СМИ, издательства
Спорт, фитнес, салоны красоты, SPA
Страхование
Строительство, проектирование, недвижимость
Сырье
Топ-персонал
Транспорт, логистика, ВЭД
Туризм, гостиницы, общественное питание
Услуги, ремонт, сервисное обслуживание
Юриспруденция


Нас интересует первая категория

In [16]:
it = catalogue.json()[0]

In [17]:
it.keys()

dict_keys(['title_rus', 'url_rus', 'title', 'title_trimmed', 'key', 'positions'])

In [18]:
for i in it['positions']: print(i['title_rus'], i['key'])

AI 651
CRM-системы 603
Data Science 627
DevOps 628
SRE 629
Web-верстка 36
Администрирование баз данных 37
Аналитика 38
Внедрение и сопровождение ПО 503
Игровое ПО / Геймдев 40
Инжиниринг 41
Интернет, создание и поддержка сайтов 42
Информационная безопасность 546
Киберспорт 620
Компьютерная анимация и мультимедиа 43
Контент 44
Мобильная разработка 604
Нейросети / Искусственный интеллект 650
Оптимизация, SEO 45
Передача данных и доступ в интернет 46
Разработка и сопровождение банковского ПО 47
Разработка, программирование 48
Сетевые технологии 49
Системная интеграция 50
Системное администрирование 51
Системы автоматизированного проектирования (САПР) 52
Системы управления предприятием (ERP) 53
Сотовые, беспроводные технологии 54
Телекоммуникации и связь 55
Тестирование, QA 56
Техническая документация 614
Техническая поддержка 57
Управление продуктом 613
Управление проектами 605
Управление разработкой 630
Юзабилити 59
Другое 60
Начало карьеры, мало опыта 61


Возьмем направления

AI 651

Data Science 627

DevOps 628

Аналитика 38

Мобильная разработка 604

Нейросети / Искусственный интеллект 650

Разработка, программирование 48

Тестирование, QA 56

Начало карьеры, мало опыта 61:

Как выяснилось, этого мало, поэтому возьмем все категории

Параметр experience возьмем из документации


In [19]:
experience = {
            "1": "без опыта",
            "2": "от 1 года",
            "3": "от 3 лет",
            "4": "от 6 лет"
        }
exp = list(map(int, experience.keys()))
exp

[1, 2, 3, 4]

In [20]:
vacancies_dict = {int(i['key']): i['title_rus'] for i in it['positions']}

In [21]:
vacancies_dict

{651: 'AI',
 603: 'CRM-системы',
 627: 'Data Science',
 628: 'DevOps',
 629: 'SRE',
 36: 'Web-верстка',
 37: 'Администрирование баз данных',
 38: 'Аналитика',
 503: 'Внедрение и сопровождение ПО',
 40: 'Игровое ПО / Геймдев',
 41: 'Инжиниринг',
 42: 'Интернет, создание и поддержка сайтов',
 546: 'Информационная безопасность',
 620: 'Киберспорт',
 43: 'Компьютерная анимация\xa0и\xa0мультимедиа',
 44: 'Контент',
 604: 'Мобильная разработка',
 650: 'Нейросети / Искусственный интеллект',
 45: 'Оптимизация, SEO',
 46: 'Передача данных и доступ в интернет',
 47: 'Разработка и сопровождение банковского ПО',
 48: 'Разработка, программирование',
 49: 'Сетевые технологии',
 50: 'Системная интеграция',
 51: 'Системное администрирование',
 52: 'Системы автоматизированного проектирования (САПР)',
 53: 'Системы управления предприятием (ERP)',
 54: 'Сотовые, беспроводные технологии',
 55: 'Телекоммуникации и связь',
 56: 'Тестирование, QA',
 614: 'Техническая документация',
 57: 'Техническая поддержк

In [22]:
vacancies_ids = list(vacancies_dict.keys())

In [23]:
vacancies_ids

[651,
 603,
 627,
 628,
 629,
 36,
 37,
 38,
 503,
 40,
 41,
 42,
 546,
 620,
 43,
 44,
 604,
 650,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 614,
 57,
 613,
 605,
 630,
 59,
 60,
 61]

In [24]:
towns.json()

{'objects': [{'id': 4,
   'id_region': 46,
   'id_country': 1,
   'title': 'Москва',
   'title_eng': 'Moscow'},
  {'id': 14,
   'id_region': 41,
   'id_country': 1,
   'title': 'Санкт-Петербург',
   'title_eng': 'Saint-Petersburg'},
  {'id': 13,
   'id_region': 50,
   'id_country': 1,
   'title': 'Новосибирск',
   'title_eng': 'Novosibirsk'},
  {'id': 33,
   'id_region': 65,
   'id_country': 1,
   'title': 'Екатеринбург',
   'title_eng': 'Ekaterinburg'},
  {'id': 12,
   'id_region': 22,
   'id_country': 1,
   'title': 'Нижний Новгород',
   'title_eng': 'Nizhnii Novgorod'},
  {'id': 73,
   'id_region': 60,
   'id_country': 1,
   'title': 'Ростов-на-Дону',
   'title_eng': 'Rostov-na-Donu'},
  {'id': 56,
   'id_region': 8,
   'id_country': 1,
   'title': 'Хабаровск',
   'title_eng': 'Khabarovsk'},
  {'id': 55,
   'id_region': 92,
   'id_country': 1,
   'title': 'Казань',
   'title_eng': "Kazan'"},
  {'id': 106,
   'id_region': 75,
   'id_country': 1,
   'title': 'Челябинск',
   'title_eng

Возьмем 5 самых крупных городов

In [25]:
town_ids = [4, 14, 13, 33, 12]

Теперь по циклу пройдемся по всем параметрам

In [26]:
import time

In [27]:
vacancies_list = []

In [28]:
params = {'town': 4,
    'post_ids': 651,
    'experience': 1,
    'page': 1,
    'count': 100,
    'catalogues': 33}
request = requests.get('https://api.superjob.ru/2.0/vacancies/',
                        headers=headers, params=params)
vacs = request.json()
vacancy_category = vacancies_dict[651]
a = vacs.get('objects', [])
for i in a:
     i['vacancy_category'] = vacancy_category
a[0]

{'canEdit': False,
 'is_closed': False,
 'id': 51760904,
 'id_client': 4944476,
 'payment_from': 0,
 'payment_to': 0,
 'date_pub_to': 1777371927,
 'date_archived': 0,
 'date_published': 1775816727,
 'address': None,
 'profession': 'Стажер инженер ИБ',
 'work': None,
 'compensation': None,
 'candidat': 'КРОК — крупная российская ИТ-компания с сильной экспертизой в области информационной безопасности. Наша команда активно растёт, и мы ищем Стажера-инженера по ИБ, готового работать с поступающими заявками и оперативно решать проблемы, возникающие в системах защиты информации.\n\nСтажировка в КРОК – это:\n\nСтарт карьеры. Стажер с первого дня становится частью команды и погружается в реальные проекты с возможностью влиять на бизнес-результат;\n\nПуть до junior-специалиста в среднем занимает 6 месяцев — все это время рядом будет наставник и команда, которые будут помогать в адаптации, делиться опытом и помогать достигать профессиональных целей.\n\nВаши задачи\n\nРабота с поступающими заявка

In [29]:
from tqdm.auto import tqdm

In [32]:
iters = len(town_ids) * len(vacancies_ids) * len(exp)
progress = tqdm(total=iters)
for t in town_ids:
    for v in vacancies_ids:
        vacancy_category = vacancies_dict[v]
        for e in exp:
            for page in range(20): #можно хоть 100, так как я настроил break
                params = {
                    'town': t,
                    'post_ids': v,
                    'experience': e,
                    'page': page,
                    'count': 100,
                    'catalogues': 33}
                try:
                    logging.info(f'отправлен запрос к апи вакансий')
                    request = requests.get('https://api.superjob.ru/2.0/vacancies/',
                                            headers=headers, params=params)
                    vacs = request.json()
                    vacs_list = vacs.get('objects', [])
                    if not vacs_list:
                        break
                    for i in vacs_list:
                        i['vacancy_category'] = vacancy_category
                        i['experience_level'] = e
                    vacancies_list += vacs_list
                    logging.info(f'успех для города {t}, категории вакансии {vacancy_category}, на странице {page}')
                except Exception as e:
                    logging.error(f'Error {e}')
                time.sleep(0.2)
            progress.update(1)

  0%|          | 0/760 [00:00<?, ?it/s]

In [33]:
df_superjob = pd.DataFrame(vacancies_list)

In [34]:
df_superjob.to_csv('superjob.csv', encoding='utf-8-sig')